### PR TITLE
Add azure cloud provider

### PR DESCRIPTION
**Pull request by feiskyer**
This is the first step of adding Azure cloud provider back (#449 and https://github.com/kubernetes/kubernetes/issues/47511). Mainly changes includes:

- bring old azure cloud provider code back
- update azure sdk to latest stable v11.1.1
- update azure sdk interfaces
- add cloud provider interface initial implementations
- unit tests

Note: Kubernetes Azure cloud provider VMSS support is not finished yet, so this work is tested togather with https://github.com/kubernetes/kubernetes/pull/55833.

And there are more work to do to enable fully azure support, which is tracking at #449.

Closes: #133

### CODE REVIEW

**Comment by feiskyer** [comment]
cc @brendandburns @mwielgus @andyzhangx

**Comment by feiskyer** [comment]
@andyzhangx Any ideas of setting up the test on Azure before VMSS is ready?

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+func testProvider(t *testing.T, m *AzureManager) *AzureCloudProvider {
+	resourceLimiter := cloudprovider.NewResourceLimiter(
+		map[string]int64{cloudprovider.ResourceNameCores: 1, cloudprovider.ResourceNameMemory: 10000000},
+		map[string]int64{cloudprovider.ResourceNameCores: 10, cloudprovider.ResourceNameMemory: 100000000})
+	provider, err := BuildAzureCloudProvider(m, nil, resourceLimiter)
+	assert.NoError(t, err)
+	return provider
+}
+
+func TestBuildAwsCloudProvider(t *testing.T) {
```
Comment (andyzhangx) [inline cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go]: Azure, not Aws

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+	provider := testProvider(t, testAzureManager)
+	err := provider.addNodeGroup("1:5:test-asg")
+	assert.NoError(t, err)
+	targetSize, err := provider.scaleSets[0].TargetSize()
+	assert.Equal(t, targetSize, 2)
+	assert.NoError(t, err)
+}
+
+func TestIncreaseSize(t *testing.T) {
+
```
Comment (andyzhangx) [inline cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go]: remove empty line

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+func TestDeleteNodes(t *testing.T) {
+	scaleSetClient := &VirtualMachineScaleSetsClientMock{}
+	m := &AzureManager{
+		scaleSets:        make([]*scaleSetInformation, 0),
+		scaleSetClient:   scaleSetClient,
+		scaleSetVmClient: &VirtualMachineScaleSetVMsClientMock{},
+		scaleSetCache:    make(map[AzureRef]*ScaleSet),
+	}
+
+	//(resourceGroupName string, vmScaleSetName string,
```
Comment (andyzhangx) [inline cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go]: remove comments

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+)
+
+// Mock for VirtualMachineScaleSetsClient
+type VirtualMachineScaleSetsClientMock struct {
+	mock.Mock
+}
+
+func (client *VirtualMachineScaleSetsClientMock) Get(resourceGroupName string,
+	vmScaleSetName string) (result compute.VirtualMachineScaleSet, err error) {
+	fmt.Printf("Called VirtualMachineScaleSetsClientMock.Get(%s,%s)\n", resourceGroupName, vmScaleSetName)
```
Comment (andyzhangx) [inline cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go]: fmt.Printf ?

**Comment by mwielgus** [comment]
Cool, thank you for this contribution! 
However, we will be able to review and merge it only after it is actually run and tested on Kubernetes.

**Comment by mwielgus** [comment]
We will also need some documentation explaining how to setup CA on Azure, as a part of this PR.

**Comment by feiskyer** [comment]
@mwielgus yep, I'm validating this on Azure now.

**Comment by feiskyer** [comment]
Validated scaling up and down on Azure VMSS (with https://github.com/kubernetes/kubernetes/pull/55833), both work as expected now. 

 @brendandburns @mwielgus @andyzhangx  Ready for review now. PTAL.

**Comment by feiskyer** [comment]
Kindly ping @MaciekPytel @mwielgus Could you take a look the PR?

**Comment by feiskyer** [comment]
ping @MaciekPytel @mwielgus Could you take a look the PR?

**Comment by mwielgus** [comment]
/lgtm

@feiskyer, @andyzhangx I assume that you will be the main contacts for all Azure-related issues and PRs. 

Neither me nor @MaciekPytel, @aleksandra-malinowska or @bskiba have capacity/capabilities to do Azure-specific reviews, debugging or testing.

**Comment by feiskyer** [comment]
@mwielgus Thanks. Yep, we will be in charge of all Azure related changes.

**Pull request merged**